src/blockchain/identity/mod.rs

```diff
@@ -0,0 +1,176 @@
+use super::Result;
+use super::{calculate_hash, Block, Blockchain};
+use crate::service::identity_service::Identity;
+use crate::util::{crypto, rsa, BcrKeys};
+use borsh::to_vec;
+use serde::{Deserialize, Serialize};
+
+#[derive(Serialize, Deserialize, Debug, Clone, PartialEq)]
+pub enum IdentityOpCode {
+    Create,
+    Update,
+    SignPersonBill,
+    CreateCompany,
+    AddSignatory,
+    RemoveSignatory,
+}
+
+#[derive(Serialize, Deserialize, Debug, Clone, PartialEq)]
+pub struct IdentityBlock {
+    pub id: u64,
+    pub hash: String,
+    pub timestamp: i64,
+    pub data: String,
+    pub public_key: String,
+    pub previous_hash: String,
+    pub signature: String,
+    pub op_code: IdentityOpCode,
+}
+
+impl Block for IdentityBlock {
+    type OpCode = IdentityOpCode;
+
+    fn id(&self) -> u64 {
+        self.id
+    }
+
+    fn timestamp(&self) -> i64 {
+        self.timestamp
+    }
+
+    fn op_code(&self) -> &Self::OpCode {
+        &self.op_code
+    }
+
+    fn hash(&self) -> &str {
+        &self.hash
+    }
+
+    fn previous_hash(&self) -> &str {
+        &self.previous_hash
+    }
+
+    fn data(&self) -> &str {
+        &self.data
+    }
+
+    fn signature(&self) -> &str {
+        &self.signature
+    }
+
+    fn public_key(&self) -> &str {
+        &self.public_key
+    }
+}
+
+impl IdentityBlock {
+    pub fn new(
+        id: u64,
+        previous_hash: String,
+        data: String,
+        op_code: IdentityOpCode,
+        keys: &BcrKeys,
+        timestamp: i64,
+    ) -> Result<Self> {
+        let hash = calculate_hash(
+            &id,
+            &previous_hash,
+            &data,
+            &timestamp,
+            &keys.get_public_key(),
+            &op_code,
+        );
+        let signature = crypto::signature(&hash, &keys.get_private_key_string())?;
+
+        Ok(Self {
+            id,
+            hash,
+            timestamp,
+            previous_hash,
+            signature,
+            public_key: keys.get_public_key(),
+            data,
+            op_code,
+        })
+    }
+}
+
+#[derive(Serialize, Deserialize, Debug, Clone)]
+pub struct IdentityBlockchain {
+    blocks: Vec<IdentityBlock>,
+}
+
+impl Blockchain for IdentityBlockchain {
+    type Block = IdentityBlock;
+
+    fn blocks(&self) -> &Vec<Self::Block> {
+        &self.blocks
+    }
+
+    fn blocks_mut(&mut self) -> &mut Vec<Self::Block> {
+        &mut self.blocks
+    }
+}
+
+impl IdentityBlockchain {
+    /// Creates a new identity chain, encrypting the identity with the public rsa key
+    pub fn new(identity: &Identity, keys: &BcrKeys, timestamp: i64) -> Result<Self> {
+        let identity_bytes = to_vec(&identity)?;
+        let genesis_hash = hex::encode(&identity_bytes);
```

Comment:
can we already use base64?

Reply:
I will add it in the next PR when I implement the operations.
I thought about an API like adding `data` and `data_encoded` and maybe even `data_decrypted` to the trait, so we can transition the bill code later easily as well.

Reply:
Hmm. Let me think about. Let's discuss tomorrow